File: Java/Pascals_Triangle.java
import java.util.*;
public class Pascals_Triangle
{
    int fact(int f)
    {
        if (f == 0)
            return 1;
        return f * fact(f - 1);
    }

    public static void main(String[] args)
    {
        int i , j;
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        Pascals_Triangle obj = new Pascals_Triangle();
        for (i = 0; i <= n; i++)
        {
            for (j = 0; j <= n - i; j++)
            {
                System.out.print(" ");//for spacing in the left
            }
            for (j = 0; j <= i; j++)
            {
                // nCr formula [n ! / ( n – r ) ! r !] 
                System.out.print(" "+ obj.fact(i)/(obj.fact(i - j)*obj.fact(j)));
            }
            System.out.println();//for changing line;
        }
        sc.close();
    }//end of main
}// end of class
/*
Sample Input & Output
The first & only line of input denotes an integer representing the no of lines
INPUT : 
5
OUTPUT :
       1
      1 1
     1 2 1
    1 3 3 1
   1 4 6 4 1
  1 5 10 10 5 1
--------------------
Time Complexity :
T(n) = O(N^2), two nested loops.
--------------------
Space complexity:
A(n) = O(N^2), for storing each of the rows of the pascal triangle.
*/

File: JavaScript/pascals_triangle.js
// Helper function to format and print the matrix in a friendly way
function printMatrix(matrix) {
  for (let i = 0; i < matrix.length; i++) {
    console.log(`[ ${matrix[i].join(' | ')} ]`);
  }
}

// Function to calculate and return pascal's triangle with given rows
function pascalsTriangle(rows) {
  let pascal = [];
  for (let i = 0; i < rows; i++) {
    let currentRow = [];
    for (let j = 0; j <= i; j++) {
      if (j === 0 || j === i) {
        // The first and last column of each row is always 1
        currentRow.push(1);
      } else {
        // The middle columns are the sum of elements of the previous row
        // The elements of the previous row (i-1) that are added are the current column (j) and the previous column (j-1)
        let temp = pascal[i - 1][j - 1] + pascal[i - 1][j];
        currentRow.push(temp);
      }
    }
    pascal.push(currentRow);
  }
  return pascal;
}

let pascal = pascalsTriangle(4);
printMatrix(pascal);

/*
Example 1:

Input
rows = 4

Output
[ 1 ]
[ 1 | 1 ]
[ 1 | 2 | 1 ]
[ 1 | 3 | 3 | 1 ]

Example 2:

Input
rows = 6

Output
[ 1 ]
[ 1 | 1 ]
[ 1 | 2 | 1 ]
[ 1 | 3 | 3 | 1 ]
[ 1 | 4 | 6 | 4 | 1 ]
[ 1 | 5 | 10 | 10 | 5 | 1 ]

TIME COMPLEXITY: O(n^2)
SPACE COMPLEXITY: O(n^2)

*/